README.txt
Library for building scrapers.

CHANGES.txt
Changelog
*********
0.7.12 (2018-03-09)
==================
* minor is_john_doe improvement

0.7.11 (2017-11-10)
==================
* minor is_person improvement

0.7.10 (2017-05-05)
==================
* minor remove_aka improvement
* minor is_person improvement

0.7.8 (2016-04-27)
==================
* minor is_person issue fixed

0.7.7 (2015-12-10)
==================
* default HTTPS handler added with unverified ssl context

0.7.6 (2015-12-01)
==================
 * proxies rotation option added
 * get rid of stand-alone readpage function
 * minor refactoring

0.7.5 (2015-11-17)
==================
 * smart_match_fullname improved

0.7.4 (2015-04-03)
==================
 * socket timeout added

0.7.3 (2015-03-12)
==================
 * AttributeError fixed

0.7.2 (2015-02-09)
==================
 * &apos; entity added
 * aka/dba/fka issue fixed

0.7.1 (2014-11-25)
==================
 * logging improved to be compatible with Splunk
 * unused imports and vars removed

0.7 (2014-10-03)
==================
 * more response info introduced to Client: resp_headers, resp_code, raw_content, real_url
 * opener.readpage changed and is not backward compatible:
    - return value type changed
    - needURL parameter removed from opener.readpage
   this doesn't impact opener.Client
 * useless class BaseCaptchaAwareClient removed
 * unused imports removed
 * minor issues with DELAY var fixed
 * opener.Client does not use X-Forwarded-For by default, which is also not backward-compatible

0.6.6 (2014-09-19)
==================
 * exceptions handling issue fixed

0.6.5 (2014-07-13)
==================
 * parse_and_normalize_streetaddress introduced to replace normalize_address
 * more non-person markers added

0.6.4 (2014-07-13)
==================
 * decaptcher_solve improvements & minor bugfix

0.6.3 (2014-07-13)
==================
 * missing optional argument added

0.6.2 (2014-06-26)
==================
 * wrong function name fixed
 
0.6.1 (2014-06-25)
==================
 * parse_ff_mapping issue fixed
 * parse_city_state_zip commas normalization added
 * decaptcher_solve - service URL can be customized to use gates like GSA
 * parse_fullname improved
 
0.6 (2014-02-01)
==================
 * Custom SSLv3 handler added to Client's opener
 * It's possible to pass custom handlers to Client

0.5.4 (2014-01-17)
==================
 * parse_city_state_zip improvements 
 * parse_fullname now understands lastname prefixes MC and VAN

0.5.3 (2013-12-12)
==================
 * smart_fullname_cmp improved. 
 * smart_match_fullname bugs fixed.

0.5.2 (2013-11-26)
==================
 * smart_fullname_cmp improved. It understands shortened forms of first and middle names
 * smart_match_fullname improved. It finds matches if a firstname or a middlename is missing.

0.5.1 (2013-10-09)
==================
 * debug flag was added for Client class. This allows to track HTTP requests & responses.
 * is_person improved

0.5.0 (2013-09-03)
==================
 * is_person improved
 * parse_fullname - work around mixed middlename & firstname when len(firstname)=1
 * DR, SIR, PHD, ESQ - new fullname suffixes, also it may come in the beginning
 * IClient, IClientFactory interfaces provided
 * Client class implements IClient interface and is IClientFactory factory
 * is_normal_address function added
 
0.4.1 (2013-07-25)
==================
 * is_person improved
 * is_defendant improved
 * is_plaintiff improved
 * walk_table now works around empty/None pages
 * parse_fullname, parse_fulladdress - strip trailing spaces, commas etc.
 * smart_fullname_cmp and smart_match_fullname added
 
0.4 (2013-04-30)
================

 * SequenceState class added
 * IConfig base interface added
 * US states & CA provinces and its codes dicts added
 * minor 'normalize' improvement
 * minor 'parse_fullname' improvement
 * minor 'parse_city_state_zip' improvement
 * MAXREADTRIES and DELAY are `readpage` optional variables and Client class attributes

0.3.13 (2013-03-25)
==================
 * remove_aka improved with FKA

0.3.12 (2013-03-07)
==================
 * is_john_doe improved
 
0.3.11 (2013-03-07)
==================
 * remove_aka improved
 * is_plaintiff, is_defendant improved
 * extract_description improved
 * parse_city_state_zip an issue with parsing data splitted by comma with no spaces fixed
 * is_person improved
 * extract_date improved
 * parse_fullname works around $ character in fullname
 * is_john_doe improved

0.3.10 (2013-03-01)
==================
 * an old bug in normalize_address was fixed

0.3.9 (2013-02-28)
==================
 * a bug in normalize_address introduced in 0.3.8 was fixed

0.3.8 (2013-02-28)
==================
 * is_person, remove_aka, normalize_address were improved

0.3.7 (2013-02-21)
==================
 * is_defendant, is_plaintiff, is_valid_attorney were improved

0.3.6 (2013-02-21)
==================
 * is_valid_attorney improved

0.3.5 (2013-02-20)
==================
 * clone_client issue when asp_state is missing fixed
 * stipping spaces from name pieces in parse_fullname
 * list of markers extended in is_person
 * is_valid_attorney added
 * smart_cmp added

0.3.4 (2013-02-18)
==================
 * missing import fixed

0.3.3 (2013-02-12)
==================
 * missing import fixed

0.3.2 (2013-02-12)
==================
 * several useful methods for US courts scraping added
 * fullname parsers were improved

0.3.1 (2013-02-08)
==================
 * minor issue with 'u\xa0' character in *normalize* fixed
 
0.3 (2013-01-25)
==================
 * headers issue in sending request fixed
 * new functions for parsing pages: get_head, get_block, get_tail, walk_table, parse_ff_mapping were added
 * function clone_client for cloning a Client object added

0.2.9 (2013-01-16)
==================
 * remove_aka and parse_fullname were improved
 * fullname parser issue fixed

0.2.8 (2013-01-14)
==================
 * fullname parser issue fixed
 
0.2.7 (2013-01-14)
==================
 * is_fullname_suffix methos added

src/sterch/scrapingtools/headers.py
### -*- coding: utf-8 -*- #############################################
# Разработано компанией Стерх (http://sterch.net/)
# Все права защищены, 2008
#
# Developed by Sterch (http://sterch.net/)
# All right reserved, 2008
#######################################################################

""" Headers dictionary

$Id: headers.py 14192 2010-04-17 08:44:49Z maxp $
"""
__author__  = "Polscha Maxim (maxp@sterch.net)"
__license__ = "<undefined>" # необходимо согласование
__version__ = "$Revision: 14192 $"
__date__ = "$Date: 2010-04-17 11:44:49 +0300 (Сб, 17 апр 2010) $"

from random import choice, randint
from copy import copy
import string

browser_headers = [
 [
                   ('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US) AppleWebKit/532.0 (KHTML, like Gecko) Chrome/3.0.195.21 Safari/532.0'),
                   ('Accept', 'application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5'),
                   ('Accept-Language', 'en-US;q=0.8,en;q=0.6'),
                   ('Accept-Charset', 'utf-8;q=0.8,latin-1;q=0,6,*;q=0.3'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US) (KHTML, like Gecko)'),
                   ('Accept', 'application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5'),
                   ('Accept-Language', 'en-US'),
                   ('Accept-Charset', 'utf-8;q=0.9,latin-1;q=0,7,*;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; uk; rv:1.9.1.3) Gecko/20090824 Firefox/3.5.3 GTB5 (.NET CLR 3.5.30729)'),
                   ('Accept', 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8,text/vnd.wap.wml;q=0.6'),
                   ('Accept-Language', 'ru,uk;q=0.8,en;q=0.7,de;q=0.5,ru-ru;q=0.3,en-us;q=0.2'),
                   ('Accept-Charset', 'utf-8;q=0.8,*;q=0.7'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; ru-RU) AppleWebKit/530.19.2 (KHTML, like Gecko) Version/4.0.2 Safari/530.19.1'),
                   ('Accept', 'application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5'),
                   ('Accept-Language', 'en-US'),
                   ('Accept-Charset', 'iso-8859-1, utf-8'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Opera/9.80 (Windows NT 5.1; U; en)'),
                   ('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1'),
                   ('Accept-Language', 'en;q=0.9,ru;q=0.1'),
                   ('Accept-Charset', 'iso-8859-1, utf-8, utf-16, *;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Opera/9.80 (Windows NT 5.1; U; en) Presto/2.2.15 Version/10.00'),
                   ('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1'),
                   ('Accept-Language', 'en-US,en;q=0.9,en;q=0.8'),
                   ('Accept-Charset', 'iso-8859-1, utf-8, utf-16, *;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; InfoPath.2; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729)'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'ru,en-US;q=0.5'),
                   ('Accept-Charset', 'iso-8859-1, utf-8, utf-16, *;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'en-US;q=0.9'),
                   ('Accept-Charset', 'iso-8859-1, utf-8, utf-16, *;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1)'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'en-US;q=0.7'),
                   ('Accept-Charset', 'iso-8859-1, utf-8, utf-16, *;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US) (KHTML, like Gecko)'),
                   ('Accept', 'application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5'),
                   ('Accept-Language', 'en-US'),
                   ('Accept-Charset', 'utf-8;q=0.8,iso-8859-1;q=0.6,*;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/5.0 (Windows; Windows NT 5.1; US; rv:1.9.1) Gecko/20090720 Firefox/3.5.1'),
                   ('Accept', 'text/html,application/xhtml+xml,application/xml;q=0.8,*/*;q=0.4,text/vnd.wap.wml;q=0.2'),
                   ('Accept-Language', 'en;q=0.7,en-us;q=0.2'),
                   ('Accept-Charset', 'utf-8, iso-8859-1;q=0.8,*;q=0.6'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/5.0 (Windows NT 5.1; en-US) AppleWebKit/530.19.2 Version/4.0.1 Safari/530.17.0'),
                   ('Accept', 'application/xml,application/xhtml+xml,text/html;q=0.8,text/plain;q=0.7,image/png,*/*;q=0.1'),
                   ('Accept-Language', 'en-US, en;q=0.9'),
                   ('Accept-Charset', 'utf-8, iso-8859-1;q=0.9'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Opera/9.85 (Windows NT 5.1; en-US)'),
                   ('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1'),
                   ('Accept-Language', 'en;q=0.9'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.5, utf-16, *;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Opera/9.88 (Windows NT 5.1; en) Presto/2.1.13 Version/9.00'),
                   ('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1'),
                   ('Accept-Language', 'en-US;q=0.9,en;q=0.3'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; .NET CLR 2.0.50700;'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'en-US;q=0.8'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.4'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/3.9)'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'en-US;q=0.9'),
                   ('Accept-Charset', 'iso-8859-1, utf-8, *;q=0.3'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1;  .NET CLR 2.0.507123;)'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'en,en-US;q=0.7'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
  [                 
                   ('User-Agent', 'Opera/9.65 (Windows; Windows NT 5.1; en) Version/9.00'),
                   ('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1'),
                   ('Accept-Language', 'en-US;q=0.9,en;q=0.3'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 7.0; Windows; Windows NT 5.1; .NET'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'en-US;q=0.9'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.4'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/3.7; .NET)'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'en-US;q=0.9'),
                   ('Accept-Charset', 'iso-8859-1, utf-8, *;q=0.3'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Mozilla/4.0 (compatible; MSIE 7.0; Windows; Windows NT 5.1; US)'),
                   ('Accept', 'image/gif, image/jpeg, image/pjpeg, image/pjpeg, application/x-shockwave-flash, application/vnd.ms-excel, application/vnd.ms-powerpoint, application/msword, application/x-silverlight, application/x-ms-application, application/x-ms-xbap, application/vnd.ms-xpsdocument, application/xaml+xml, */*'),
                   ('Accept-Language', 'en,en-US;q=0.7'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Opera/9.68 (Windows; Windows; Windows NT 5.1; en)'),
                   ('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1'),
                   ('Accept-Language', 'en-US;q=0.9,en;q=0.3'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Opera/9.68 ( Windows NT 5.1; en) Version/9.00'),
                   ('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1'),
                   ('Accept-Language', 'en-US;q=0.9,en;q=0.3'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
 [                 
                   ('User-Agent', 'Opera/9.67 (Windows; Windows NT 5.1; en) Version/9.00'),
                   ('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1'),
                   ('Accept-Language', 'en-US;q=0.9,en;q=0.3'),
                   ('Accept-Charset', 'iso-8859-1, utf-8;q=0.1'),
                   ('Keep-Alive', '300'),
                   ('Connection', 'keep-alive')
 ],
]

def getheaders():
    """ Returns random browser headers """
    headers = copy(choice(browser_headers))
    ua, value = headers[0]
    name = "".join([choice(string.letters) for l in xrange(0,randint(3,10))])
    major = str(randint(0,9))
    middle =  str(randint(0,20))
    minor =  str(randint(0,150))
    tail = "; %s v%s.%s.%s;" % (name, major, middle, minor)
    index = value.find(")")
    if index > 0:
        value = value[:index] + tail + value[index:]  
    else:
        value = value + tail
    headers[0] = (ua, value)
    headers.append(("X-Forwarded-For", ".".join(map(str,[10,randint(0,253),randint(0,253),randint(1,253)]))))
    
    return headers 